crates/rooch-framework/sources/auth_validator/ecdsa_k1_validator.move

```diff
@@ -83,61 +82,95 @@ module rooch_framework::ecdsa_k1_validator {
-            option::extract(&mut auth_key_option)
+            auth_key_option
         }else {
-            //if AuthenticationKey does not exist, return addr as authentication key
-            moveos_std::bcs::to_bytes(&addr)
+            option::none<vector<u8>>()
         }
     }
 
-    public fun validate(ctx: &StorageContext, authenticator_payload: vector<u8>) {
-        // TODO handle non-ed25519 auth key and address relationship
-        // let auth_key = ecdsa_k1_authentication_key(&authenticator_payload);
-        // let auth_key_in_account = get_authentication_key(ctx, storage_context::sender(ctx));
-        // assert!(
-        //    auth_key_in_account == auth_key,
-        //    auth_validator::error_invalid_account_auth_key()
-        // );
+    /// The authentication key exists in account or not.
+    public fun is_authentication_key_in_account(ctx: &StorageContext, addr: address): bool {
+        option::is_some(&get_authentication_key_option_from_account(ctx, addr))
+    }
+
+    /// Get the authentication key of the given account.
+    public fun get_authentication_key_from_account(ctx: &StorageContext, addr: address): vector<u8> {
+        assert!(is_authentication_key_in_account(ctx, addr), error::not_found(EAuthenticationKeyNotFoundInAccount));
+        option::extract(&mut get_authentication_key_option_from_account(ctx, addr))
+    }
+
+    /// Only validate the authenticator's signature.
+    public fun validate_signature(authenticator_payload: &vector<u8>, tx_hash: &vector<u8>) {
         assert!(
             ecdsa_k1::verify(
-                &ecdsa_k1_signature(&authenticator_payload),
-                &ecdsa_k1_public_key(&authenticator_payload),
-                &storage_context::tx_hash(ctx),
-                SHA256, // KECCAK256:0, SHA256:1, TODO: The hash type may need to be passed through the authenticator
+                &get_signature_from_authenticator_payload(authenticator_payload),
+                &get_public_key_from_authenticator_payload(authenticator_payload),
+                tx_hash,
+                SHA256
             ),
             auth_validator::error_invalid_authenticator()
         );
     }
 
+    public fun validate(ctx: &StorageContext, authenticator_payload: vector<u8>) {
+        let tx_hash = storage_context::tx_hash(ctx);
+        validate_signature(&authenticator_payload, &tx_hash);
+
+        let auth_key_from_authenticator_payload = get_authentication_key_from_authenticator_payload(&authenticator_payload);
+        std::debug::print(&auth_key_from_authenticator_payload);
+        // Although we have checked public key length in rotate_authentication_key_entry function,
+        // it needs to validate the authentication key isn't empty or malformed.
+        assert!(
+           vector::length(&auth_key_from_authenticator_payload) == V_AUTHENTICATION_KEY_LENGTH,
+           error::invalid_argument(EInvalidAuthenticatorPayloadAuthKeyLength)
+        );
```

Comment:
We need to compare the auth_key from the payload with the auth_key from the account.
if the auth_key from the account is none, and the MultiChainAddress in context is not empty, we can use the ETH address as the auth_key.

Reply:
Did you mean context the TxContext?

```zsh
[debug] 0x2::storage_context::StorageContext {
  tx_context: 0x2::tx_context::TxContext {
    sender: @0xebf29d2aed4da3d2e13a32d71266a302fbfd5ceb3ff1f465c006fa207f1789ce,
    tx_hash: 0x6ab9102e46eccda49488d033219b1ba9992f446608115f72061a70e0313e5c81,
    ids_created: 0,
    map: 0x2::simple_map::SimpleMap<0x1::string::String, 0x2::copyable_any::Any> {
      data: []
    }
  },
  object_storage: 0x2::object_storage::ObjectStorage {
    handle: 0x2::object_id::ObjectID {
      id: @0x0
    }
  }
}
```

Unfortunately, I couldn't find for MultiChainAddress, and the sender in the context, the auth_key from the account, is default to none for non-ed25519 schemes since only ed25519 address exists at the beginning and I need to call rotate auth keys to update a one in the validator store.

Reply:
If the user uses MetaMask and submits the Ethereum transaction, the MultiChainAddress will in the TxContext

https://github.com/rooch-network/rooch/blob/420a92379b226777689fce8245c0e6cdb7405c79/crates/rooch-framework/sources/transaction_validator.move#L96-L104

Reply:
> If the user uses MetaMask and submits the Ethereum transaction, the MultiChainAddress will in the TxContext
> 
> https://github.com/rooch-network/rooch/blob/420a92379b226777689fce8245c0e6cdb7405c79/crates/rooch-framework/sources/transaction_validator.move#L96-L104

I've sent a transaction from MetaMask and it seems it has successfully included a map data:

```zsh
[debug] 0x2::storage_context::StorageContext {
  tx_context: 0x2::tx_context::TxContext {
    sender: @0x5920595edf45d7e863493a3f9182b4101dd2bc27146fce813081c087d3e1d3b9,
    tx_hash: 0x2928a5913dfe71aa9a809238e9d7235b608c02d29ce625abc9f970836c94f0fd,
    ids_created: 0,
    map: 0x2::simple_map::SimpleMap<0x1::string::String, 0x2::copyable_any::Any> {
      data: [
        0x2::simple_map::Element<0x1::string::String, 0x2::copyable_any::Any> {
          key: "0000000000000000000000000000000000000000000000000000000000000002::tx_meta::TxMeta",
          value: 0x2::copyable_any::Any {
            type_name: "0000000000000000000000000000000000000000000000000000000000000002::tx_meta::TxMeta",
            data: 0x010195b15595f939a8cf5ece25d44fc832acaf218a6028472b9c0785338fae21bef907636f756e74657208696e637265617365
          }
        }
      ]
    }
  },
  object_storage: 0x2::object_storage::ObjectStorage {
    handle: 0x2::object_id::ObjectID {
      id: @0x0
    }
  }
}
```

Anyway can I verify the `TxMeta`'s data `0x010195b15595f939a8cf5ece25d44fc832acaf218a6028472b9c0785338fae21bef907636f756e74657208696e637265617365` is indeed the MetaMask sender's address?

Reply:
@jolestar I used MetaMask to connect to a localhost Rooch RPC, but it seems that there's no multi-chain address mapping in the `tx_context`?